src/messaging/ExchangeContext.cpp

```diff
@@ -444,7 +444,14 @@ void ExchangeContext::OnSessionReleased()
 
 CHIP_ERROR ExchangeContext::StartResponseTimer()
 {
-    System::Layer * lSystemLayer = mExchangeMgr->GetSessionManager()->SystemLayer();
+    // Verify that the Session Manager is still alive. Note that host applications may stop the chip::Server
+    // at any time, which will tear down the Session Manager, Exchange Manager, and Exchange Context. Note
```

Comment:
No, they're not allowed to stop it at any time, unless they are holding the Matter stack lock.  We should add an assert to that effect.

Reply:
Can you share any references to this expectation in the codebase? It isn't enforced anywhere in `ExchangeMgr` or any other file in `messaging/*` except for a single usage in `ExchangeContext` at the beginning of `SendMessage`.

Reply:
And note that the lock asserts are no-ops on various platforms.  Again, the expectation is that SDK consumers do the right thing and only touch SDK APIs with the locks held, either directly (which _also_ does not work on some lock-free platforms) or via posting whatever they are doing to the Matter event loop.

Reply:
>Which expectation? That you only touch the Matter SDK while holding the Matter lock? It's a basic API expectation in all Matter APIs.

It's surprising that this assertion would not be mentioned in all but one file in this module. As a newcomer, and having only read a few dozen files in the SDK, this is the only time that I've seen it mentioned. Is there a centralized document explaining this thread safety approach?

>Again, the expectation is that SDK consumers do the right thing

This doesn't seem like a reasonable assumption for an SDK... public APIs should generally be defensive and assume that the caller can always make the wrong decision. Even if the SDK chooses not to fail gracefully (a valid choice), clients need immediate feedback to help them find bugs in their code. The current segmentation fault in the middle of `SendMessage` being serviced by a different thread doesn't obviously indicate that the client (the `tv-casting-app`) is doing something wrong.

Reply:
>  Is there a centralized document explaining this thread safety approach?

I don't know of one... @mrjerryjohns did something ever get written here?

To be clear, this is not great, but that's the state of things right now.  Not enough hands....

> public APIs should generally be defensive and assume that the caller can always make the wrong decision. 

Sure.  Again:

1. Not all platforms support checking the locking correctness.
2. Adding the asserts to all the places that should have them has nonzero cost both in developer time and final codesize on the platforms that do have such checking.
3. We do not have a clear distinction between private and public APIs right now (an issue in its own right).

We have been adding the asserts in an ad-hoc manner to high-value places that would have a good chance of catching bugs, but something more systematic would be welcome, obviously.

> The current segmentation fault in the middle of SendMessage being serviced by a different thread doesn't obviously indicate that the client (the tv-casting-app) is doing something wrong.

Completely agreed!  I'm just saying that the right fix here is not in SendMessage; that's going to cover up a thread race that's likely corrupting data elsewhere, at least probabilistically.  The right fix is to get rid of the thread race.

Reply:
Thanks for the good discussion on this issue, I agree that we should fix the thread race instead. With #25041 merged, we should start seeing assertion failures now in the `tv-casting-app` example which would help pinpoint the threading problem. I'm happy to close out this PR, I'll just wait a few days to see if anyone chimes in with pointers on the thread safety documentation. Thanks!